File: AWS/iac-microbenchmark/ec2/typescript/ecc384_sign.ts
#!/usr/bin/env node
import { KMSClient, SignCommand } from '@aws-sdk/client-kms';
import * as process from 'process';
import * as base64 from 'base64-js';

// Initialize the AWS KMS client with the specified region
const kmsClient = new KMSClient({
  region: 'us-east-1' // Specify the desired region
});

async function main(): Promise<void> {
  try {
    // Get the JSON input from command line arguments
    const requestJsonRaw: string = process.argv[2];

    // Parse the JSON input
    const requestJson = JSON.parse(requestJsonRaw);

    // Extract the message from the parsed JSON
    const message: string = requestJson.get('message');

    // Retrieve the KMS key ARN from environment variables
    const eccKmsKeyId = process.env['ECC384_KMS_KEY_ARN'];

    if (!eccKmsKeyId) {
      throw new Error("ECC384_KMS_KEY_ARN environment variable is not set");
    }

    // Sign the message using KMS
    const signCommand = new SignCommand({
      KeyId: eccKmsKeyId,
      Message: Buffer.from(message, 'utf-8'),
      MessageType: 'RAW',
      SigningAlgorithm: 'ECDSA_SHA_384' // Use 'ECDSA_SHA_384' for P-384
    });

    const response = await kmsClient.send(signCommand);
    const signature = response.Signature as Uint8Array;

    // Encode the signature to base64 for easier transport
    const signatureB64 = base64.fromByteArray(signature);

    // Output the result as a JSON string
    console.log(JSON.stringify({ signature: signatureB64 }));
  } catch (error) {
    // Handle and log any errors that occur
    console.error("Error:", error);
  }
}

// Execute the main function
main();

File: AWS/iac-microbenchmark/ec2/python/ecc384_sign.py
#!/usr/bin/env python3
import json
import boto3
import base64
from botocore.exceptions import ClientError
import os
import sys

# Initialize the KMS client
kms_client = boto3.client('kms')

def main():

    # Get the name argument from sys.argv
    request_json_raw = sys.argv[1]

    request_json = json.loads(request_json_raw)

    # Extract the message from the event payload
    message = request_json.get('message')

    ecc_kms_key_id = os.environ['ECC384_KMS_KEY_ARN']
    # Sign the message
    response = kms_client.sign(
        KeyId=ecc_kms_key_id,
        Message=message.encode('utf-8'),
        MessageType='RAW',
        SigningAlgorithm='ECDSA_SHA_384'  # Use 'ECDSA_SHA_384' for P-384
    )
    
    signature = response['Signature']

    # Encode the signature to base64 for easier transport
    signature_b64 = base64.b64encode(signature).decode('utf-8')
    print(json.dumps({'signature': signature_b64}))

if __name__ == "__main__":
    main()
